File: src/WorldArea.cpp
#include "WorldArea.h"
#include "Grid.h"
#include "color.h"

#include <boglfw/renderOpenGL/Shape2D.h>

WorldArea::WorldArea(Grid *g, GridPoint topLeft, GridPoint bottomRight)
	: grid_(g)
	, topLeft_(topLeft), bottomRight_(bottomRight) {
}

WorldArea::~WorldArea() {
}

void WorldArea::draw(Viewport*) {
	ScreenPoint sTopLeft = grid_->gridToScreen(topLeft_);
	ScreenPoint sBottomRight = grid_->gridToScreen(bottomRight_);

	const int thickness = 1;

	// left edge
	glm::vec2 pos {sTopLeft.x - thickness, sTopLeft.y - thickness};
	glm::vec2 sz {1 + 2*thickness, (sBottomRight.y - sTopLeft.y) + 2*thickness};
	Shape2D::get()->drawRectangleFilled(pos, 0, sz, Colors::GRID);
	// top edge
	pos = {sTopLeft.x - thickness, sTopLeft.y - thickness};
	sz = {(sBottomRight.x - sTopLeft.x) + 2*thickness, 1 + 2*thickness};
	Shape2D::get()->drawRectangleFilled(pos, 0, sz, Colors::GRID);
	// right edge
	pos = {sBottomRight.x - thickness, sTopLeft.y - thickness};
	sz = {1 + 2*thickness, (sBottomRight.y - sTopLeft.y) + 2*thickness};
	Shape2D::get()->drawRectangleFilled(pos, 0, sz, Colors::GRID);
	// bottom edge
	pos = {sTopLeft.x - thickness, sBottomRight.y - thickness};
	sz = {(sBottomRight.x - sTopLeft.x) + 2*thickness, 1 + 2*thickness};
	Shape2D::get()->drawRectangleFilled(pos, 0, sz, Colors::GRID);
}

bool WorldArea::containsPoint(WorldPoint const& wp) const {
	WorldPoint wTopLeft = grid_->gridToWorld(topLeft_);
	WorldPoint wBottomRight = grid_->gridToWorld(bottomRight_);
	return wp.x > wTopLeft.x && wp.y > wTopLeft.y
		&& wp.x < wBottomRight.x && wp.y < wBottomRight.y;
}

File: client/src/world-area.ts
import { Grid } from "./grid";
import { GridPoint, ScreenPoint, WorldPoint } from "./math/math";
import { Vector } from "./math/vector";

export class WorldArea {
	constructor(
		private readonly grid_: Grid,
		private readonly topLeft_: GridPoint,
		private readonly bottomRight_: GridPoint,
	) {}

	draw(): void {
		const sTopLeft: ScreenPoint = this.grid_.gridToScreen(this.topLeft_);
		const sBottomRight: ScreenPoint = this.grid_.gridToScreen(this.bottomRight_);

		const thickness = 1;

		// left edge
		let pos = new Vector(sTopLeft.x - thickness, sTopLeft.y - thickness);
		let sz = new Vector(1 + 2 * thickness, sBottomRight.y - sTopLeft.y + 2 * thickness);
		// FIXME draw
		//Shape2D::get().drawRectangleFilled(pos, 0, sz, Colors::GRID);

		// top edge
		pos = new Vector(sTopLeft.x - thickness, sTopLeft.y - thickness);
		sz = new Vector(sBottomRight.x - sTopLeft.x + 2 * thickness, 1 + 2 * thickness);
		// FIXME draw
		//Shape2D::get().drawRectangleFilled(pos, 0, sz, Colors::GRID);

		// right edge
		pos = new Vector(sBottomRight.x - thickness, sTopLeft.y - thickness);
		sz = new Vector(1 + 2 * thickness, sBottomRight.y - sTopLeft.y + 2 * thickness);
		// FIXME draw
		//Shape2D::get().drawRectangleFilled(pos, 0, sz, Colors::GRID);

		// bottom edge
		pos = new Vector(sTopLeft.x - thickness, sBottomRight.y - thickness);
		sz = new Vector(sBottomRight.x - sTopLeft.x + 2 * thickness, 1 + 2 * thickness);
		// FIXME draw
		// Shape2D::get().drawRectangleFilled(pos, 0, sz, Colors::GRID);
	}

	containsPoint(wp: WorldPoint): boolean {
		const wTopLeft: WorldPoint = this.grid_.gridToWorld(this.topLeft_);
		const wBottomRight: WorldPoint = this.grid_.gridToWorld(this.bottomRight_);
		return wp.x > wTopLeft.x && wp.y > wTopLeft.y && wp.x < wBottomRight.x && wp.y < wBottomRight.y;
	}

	topLeft(): GridPoint {
		return this.topLeft_;
	}
	bottomRight(): GridPoint {
		return this.bottomRight_;
	}
}
